src/evo/creditpool.cpp

```diff
@@ -0,0 +1,344 @@
+// Copyright (c) 2023 The Dash Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <evo/creditpool.h>
+
+#include <evo/assetlocktx.h>
+#include <evo/cbtx.h>
+
+#include <llmq/utils.h>
+
+#include <chain.h>
+#include <logging.h>
+#include <validation.h>
+
+#include <algorithm>
+#include <exception>
+#include <memory>
+#include <stack>
+
+static const std::string DB_CREDITPOOL_SNAPSHOT = "cpm_S";
+
+std::unique_ptr<CCreditPoolManager> creditPoolManager;
+
+static bool getDataFromUnlockTx(const CTransaction& tx, CAmount& toUnlock, uint64_t& index, TxValidationState& state)
+{
+    CAssetUnlockPayload assetUnlockTx;
+    if (!GetTxPayload(tx, assetUnlockTx)) {
+        return state.Invalid(TxValidationResult::TX_CONSENSUS, "failed-creditpool-unlock-payload");
+    }
+
+    index = assetUnlockTx.getIndex();
+    toUnlock = assetUnlockTx.getFee();
+    for (const CTxOut& txout : tx.vout) {
+        if (!MoneyRange(txout.nValue)) {
+            return state.Invalid(TxValidationResult::TX_CONSENSUS, "failed-creditpool-unlock-txout-outofrange");
+        }
+        toUnlock += txout.nValue;
+    }
+    return true;
+}
+
+namespace {
+    struct UnlockDataPerBlock  {
+        CAmount unlocked{0};
+        std::unordered_set<uint64_t> indexes;
+    };
+} // anonymous namespace
+
+// it throws exception if anything went wrong
+static UnlockDataPerBlock getDataFromUnlockTxes(const std::vector<CTransactionRef>& vtx)
+{
+    UnlockDataPerBlock blockData;
+
+    for (CTransactionRef tx : vtx) {
+        if (tx->nVersion != 3 || tx->nType != TRANSACTION_ASSET_UNLOCK) continue;
+
+        CAmount unlocked{0};
+        TxValidationState tx_state;
+        uint64_t index{0};
+        if (!getDataFromUnlockTx(*tx, unlocked, index, tx_state)) {
+            throw std::runtime_error(strprintf("%s: CCreditPoolManager::getCreditPool failed: %s", __func__, tx_state.ToString()));
+        }
+        blockData.unlocked += unlocked;
+        blockData.indexes.insert(index);
+    }
+    return blockData;
+}
+
+bool CSkipSet::add(uint64_t value)
+{
+    assert(!contains(value));
+
+    if (auto it = skipped.find(value); it != skipped.end()) {
+        skipped.erase(it);
+        return true;
+    }
+
+    assert(current_max <= value);
+    if (capacity() + value - current_max > capacity_limit) {
+        LogPrintf("CSkipSet::add failed due to capacity exceeded: requested %lld to %lld while limit is %lld\n",
+                value - current_max, capacity(), capacity_limit);
+        return false;
+    }
+    for (uint64_t index = current_max; index < value; ++index) {
+        bool insert_ret = skipped.insert(index).second;
+        assert(insert_ret);
+    }
+    current_max = value + 1;
+    return true;
+}
+
+bool CSkipSet::canBeAdded(uint64_t value) const
+{
+    if (contains(value)) return false;
+
+    if (skipped.find(value) != skipped.end()) return true;
```

Comment:
not exactly. If we will add element in the middle (when this value is already skipped!) in this case no extra memory will be used, more over memory usage will be reduced. So, even if there's no any extra space, the value may be added.

On other hand, in this condition:
`if (capacity() + value - current_max > capacity_limit) {`
Let's assume that capacity is 1, value is 1, current_max is 20.
data structure have all numbers from 2 to 20. Internal presentation is: current_max = 20, skipped: {1}.
Let's try to add value "1": `1 + 1 - 20 = negative number`  and comparision will fail (because all numbers here unsigned).

So, value can't be added.
-------------------------
@UdjinM6 
Maybe replace here "skipping list" structure to "ranges"? such as [2, 20), [21, 45), [50, 51). It's easy to provide auto-merging ranges when elements are added; performance is similar.
**Pros:** Adding operations should not care if there's somehow `index` with big gap - it won't eat ram accidentally
**Cons:** eats more memories in case if gaps are small (1 number only) - consume memory for 2 number (left/right)  instead 1 number.

also it won't stuck all asset-unlocks txes once if capacity will run out - it will just eat little more memory (linear from amount of gaps, but independent how big they are).